docker-compose.example.yml

```
@@ -0,0 +1,34 @@
+services:
+
+  app:
+    image: ghcr.io/maybe-finance/maybe:latest
+    ports:
+      - 3000:3000
```

I believe this should be:
```suggestion
      - "127.0.0.1:3000:3000"
```
so the Rails backend is not exposed to the internet without a reverse proxy in front of it

---

I don't think that's needed, and even then if you want to stop it being exposed to the host (not internet, since that would require port-forwarding) just don't bind the ports and use a reverse proxy container.

---

>being exposed to the host (not internet, since that would require port-forwarding)

@gantoine I think your vision of how this works is incorrect, especially if you're talking about servers that are exposed to the internet directly.

`3000:3000` is equivalent to `0.0.0.0:3000:3000` which means `listen on port 3000 on ALL interfaces` so if your server has an interface that's exposed to the internet (literally any VPS) then it will accept incoming connections on port 3000 from virtually anyone. That's not a safe/sane default.

> just don't bind the ports and use a reverse proxy container

That's another assumption that's going to make it harder for anyone interested in self-hosting, since your reverse proxy container will listen on ports 80 and 443, and you will have to put any/all subsequent containers in the same Docker network and configure the reverse proxy serving your Maybe instance to serve your other containers or use some convoluted setup involving Traefik and labeling stuff.

I, for example, have a Caddy on the host doing all the reverse proxying to all containers I have on my server like Immich, Maybe, Nextcloud and a ton of other stuff.

---

Just to confirm—you're optimizing for a self-hosting scenario here where you're running this app on your _local_ machine and _never_ want it exposed to the internet?

---

@zachgoll Me? No, I'm optimizing for any sane self-hosting solution - there's a reverse proxy in front of the Rails application. If we put a reverse proxy in the Docker Compose configuration and the self-hosting user has some other services on the server, then they have two options.

1. Modify the provided `docker-compose.yml` so the reverse proxy doesn't listen on ports 80 and 443 and point their preferred reverse proxy to... reverse proxy to our reverse proxy
2. Dump all their other containerized services into the same Docker network and reconfigure our reverse proxy to work with their services.

Maybe there's some convoluted setup leveraging Treafik, but that's not we are aiming for

---

What we *could* do is have an additional `docker-compose.override.yml` sample which has a reverse proxy configured inside. Then we can point everyone who doesn't particularly care about those details to use alongside the default `docker-compose.yml`

WDYT?

---

@Quintasan Your comment is perfectly reasonable, it's just that I've never seen an example docker-compose bind the network interface (in like, any self-hosted software I use or have tried).

> literally any VPS

I'm not thinking about the VPS case, I'm more-so thinking about the user who's first getting started with self hosting on a small server/old laptop/RPI at home.

> you will have to put any/all subsequent containers in the same Docker network and configure the reverse proxy serving your Maybe instance to serve your other containers

That's how I have it setup at home, and it works well pretty well 🤷🏼 

> you're optimizing for a self-hosting scenario here where you're running this app on your local machine and never want it exposed to the internet?

A machine running in my home, accessible via tailscale or cloudflare tunnel.

---

@Quintasan @gantoine thanks for this additional detail, this is helpful.

While I think both of you are probably stronger than me on the config details here, here's the general direction I was thinking:

- `docker-compose.example.yml` - this should represent a setup that intersects the 1) most common and 2) most "beginner friendly" setup.  It should _nearly_ be an "out of the box" copy/paste solution, or as close as we can get.
- `docker-compose.[specific-setup].example.yml` - we could introduce variations of the default as sample files OR add instructions to our documentation

---

I’m just a self-hosting bystander here, but I would tend to agree with @gantoine. I’ve never seen Docker compose set ups specifically bind the network interface. And, even on a VPS set up, you would need to specifically allow that port in the firewall; at least that’s how it has been for any fresh box I have installed.

I also think if most people are using a Docker set up, there’s no real need to “dumb it down”. They should already understand the risks with simple networking concepts like ports and how to securely access your apps over the internet. I think the Render deploy option serves that purpose well.

---

A very, very simplified example:

```
# docker-compose.yml
# Ours will be slightly more verbose
services:
  backend:
    image: traefik/whoami
    command:
      - --port=6666
    ports:
      - "127.0.0.1:6666:6666"
```

```
# docker-compose.override.yml
services:
  caddy:
    image: caddy:alpine
    command: caddy reverse-proxy --from <domain to serve on> --to backend:6666
    ports:
      - 80:80
      - 443:443
```

If you `docker-compose up -d` with those two files, then you end up with a backend container and a reverse proxy for it with automatic HTTPS. We can supply a custom Caddyfile OR use Traefik as the reverse proxy if we need a more complicated setup

---

>I’ve never seen Docker compose set ups specifically bind the network interface. And, even on a VPS set up, you would need to specifically allow that port in the firewall; at least that’s how it has been for any fresh box I have installed.

All OVH and Hetzner boxes I have had so far accept all connections on all ports, unless you explicitly configure them not to do so. Assuming your boxes don't (which is a good thing), binding to `127.0.0.1:<port>` changes exactly nothing in your case, while being a sane default for all boxes which don't have a firewall. I remain unconvinced that binding to `0.0.0.0` and exposing the backend container directly to the internet is a good idea.

> I also think if most people are using a Docker set up, there’s no real need to “dumb it down”.

I don't think we're dumbing anything down here. We're trying to provide sane and secure defaults while giving people option to `just run it` without any tinkering. If someone wants to tinker with this setup, then they have all the escape hatches (`docker-compose.override.yml` and the ability to bring their own reverse proxy)

---

> binding to 127.0.0.1:<port> changes exactly nothing in your case, while being a sane default for all boxes which don't have a firewall.

This is a fair point, and as it personally wouldn't affect me (as I know how to edit a compose file) I think it's fine to set it that way in the default. 👍🏼

---

I think @Quintasan's proposal here makes sense overall:

1. Give a sane default that "just works"
2. Provide escape hatches for those with the knowledge and/or motivation to customize